### ISSUE TITLE
Warn on unused cast result

### ISSUE THREAD

**Issue by KotlinIsland** [python/mypy#12059]
```py
from typing import cast

def foo(a: object) -> None:
    cast(int, a)
    print(a + 1)
```
This should be reported imo, the cast does nothing at all here is is most likely a mistake.

This is a common pattern in other languages:
```kt
fun foo(a: Any) {
    a as Int
    print(a + 1)
}
```

I think people would accidentally attempt this pattern, expecting it to narrow the type of the variable, but mypy doesn't work that way.

**Comment by erictraut** [comment]
This is a legitimate upcast. Why do you think it's useless?

Is your point that the the return result of the cast call isn't assigned to a variable or otherwise used? That's not a problem with the cast call itself, and it's probably not the job of a type checker to flag it as a problem. It's more of a linter feature.

Or was your intent to provide an example where the cast is truly unnecessary (redundant) from a type standpoint, like in the example below? Mypy already has an option for that called "warn-redundant-casts".

```python
def foo(a: str) -> None:
    x = cast(str, a)
```

**Comment by KotlinIsland** [comment]
The issue is that a generic linter wouldn't be able to detect this is a redundant call, there is no way to annotate a function as pure and otherwise a linter would have to special case this function.

Seeing as this is a typing related function I would think it falls into mypys boat to warn about, this is most likely incorrect code, but a common pattern in other languages:
```kt
fun foo(a: Any) {
  a as Int
  print(a + 1)
}
```

For these reasons I think it would be fitting for mypy to warn in this scenario.

**Comment by JelleZijlstra** [comment]
So what you'd want is an equivalent to GCC's `__attribute__ ((warn_unused_result))` (https://gcc.gnu.org/onlinedocs/gcc-4.7.2/gcc/Function-Attributes.html). I think that's been brought up before, but it's better discussed on the typing tracker.

**Comment by KotlinIsland** [comment]
Well in this case a `pure` annotation would be good, I also like the idea of a must use result annotation.

But for this instance can it just be special cased in mypy?

**Comment by KotlinIsland** [comment]
> This is a legitimate upcast

`object` cast to `str` is a downcast.

**Comment by erictraut** [comment]
Yep, it is a downcast. Good catch.

**Comment by AlexWaygood** [comment]
> Well in this case a `pure` decorator would be good, Idea warns on unused return value of pure functions.

This feels unfeasible: the number of functions in typeshed that would have to be decorated with `@pure` would be vast.

**Comment by KotlinIsland** [comment]
All this talk of decorators is way off topic 😳.

Mypy loves to special case typing centric functions, and as outlined above I believe this is a good candidate.

**Comment by AlexWaygood** [comment]
> All this talk of decorators is way off topic 😳.
> 
> Mypy loves to special case typing centric functions, and as outlined above I believe this is a good candidate.

You were the one who brought up a decorator! But, fair enough :)

**Comment by hauntsaninja** [comment]
This has been open for a while and no one has special cased it, closing as a duplicate of https://github.com/python/mypy/issues/6936
